### PR TITLE
Output modes for classification added

### PR DESCRIPTION
**Pull request by nicl-nno**
In the current version only probabilities can be returned for the classification problems #187 .

Now the 3 mode is implemented:
'probs'/'default' - probability of positive class (all probabilities if >2 classes)
'full_probs' - probabilities for each class
'labels' - labels of the classes (0,1,2...)

### CODE REVIEW

File: fedot/core/composer/chain.py [hunk 1]
```diff
@@ -86,19 +86,21 @@ def fit(self, input_data: InputData, use_cache=True, verbose=False):
         train_predicted = self.root_node.fit(input_data=input_data, verbose=verbose)
         return train_predicted
 
-    def predict(self, input_data: InputData):
+    def predict(self, input_data: InputData, output_mode: str = 'default'):
         """
         Run the predict process in all nodes in chain starting with root.
 
         :param input_data: data for prediction
+        :param output_mode: desired output for models (e.g. labels, probs, full_probs)
```
Comment (J3FALL) [inline fedot/core/composer/chain.py:94]: 1. Там еще есть опция default, думаю, стоит про нее указать, что является по умолчанию.
2. Чем отличается probs от full_probs? 
full_probs - это когда выдаются вероятности каждого класса, а probs - только лучшего, или как?

File: fedot/core/composer/node.py [hunk 2]
```diff
@@ -121,11 +119,12 @@ def fit(self, input_data: InputData, verbose=False) -> OutputData:
 
         return self.output_from_prediction(input_data, model_predict)
 
-    def predict(self, input_data: InputData, verbose=False) -> OutputData:
+    def predict(self, input_data: InputData, output_mode: str = 'default', verbose=False) -> OutputData:
```
Comment (J3FALL) [inline fedot/core/composer/node.py:122]: Добавь здесь перенос строки.

File: fedot/core/models/evaluation/evaluation.py [hunk 3]
```diff
@@ -211,12 +213,16 @@ def predict(self, trained_model, predict_data: InputData):
...
-            raise NotImplementedError()
-        elif n_classes == 2:
-            prediction = prediction[:, 1]
-
+        if self.output_mode == 'labels':
+            prediction = trained_model.predict(predict_data.features)
+        elif self.output_mode in ['probs', 'full_probs', 'default']:
+            prediction = trained_model.predict_proba(predict_data.features)
+            if n_classes < 2:
+                raise NotImplementedError()
```
Comment (J3FALL) [inline fedot/core/models/evaluation/evaluation.py:221]: Вот этот кейс не очень понял.

File: fedot/core/models/evaluation/evaluation.py [hunk 4]
```diff
@@ -24,7 +24,7 @@
 from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
```
Comment (J3FALL) [inline fedot/core/models/evaluation/evaluation.py:24]: Смотри, а если в цепочке будет не sklearn-модель?
По логике получается, что и для моделей из других фреймворков нужна эта штука.
Как отработает скрипт, если в цепочке будет, например, нейронка из keras-а, а ей задать недефолтный output_mode?

File: test/tasks/test_classification.py [hunk 5]
```diff
@@ -80,3 +90,31 @@ def test_classification_with_pca_chain_fit_correct():
...
+    data = get_binary_classification_data()
+    chain = chain_simple()
+    train_data, test_data = train_test_data_setup(data, shuffle_flag=True)
+
+    chain.fit(input_data=train_data)
+    results = chain.predict(input_data=test_data, output_mode='full_probs')
+    results_probs = chain.predict(input_data=test_data)
+
+    assert not np.array_equal(results_probs.predict, results.predict)
+    assert results.predict.shape == (len(test_data.target), 2)
```
Comment (J3FALL) [inline test/tasks/test_classification.py:120]: Уже спрашивал выше про output_mode=probs.
Тут нет теста на этот пример, почему?

**Review by J3FALL — CHANGES_REQUESTED** [review]

File: fedot/core/models/evaluation/evaluation.py (hunk 3 above)
Comment (nicl-nno) [reply fedot/core/models/evaluation/evaluation.py:221]: Если нет хотя бы двух классов в данных, то что-то пошло не так и классификация не имеет смысла.

File: fedot/core/models/evaluation/evaluation.py (hunk 4 above)
Comment (nicl-nno) [reply fedot/core/models/evaluation/evaluation.py:24]: Да, придется делать так, но сейчас у нас все модели-классификаторы вроде используетю sklearn-стратегию. Keras сейчас только для рядов.

File: fedot/core/models/evaluation/evaluation.py (hunk 3 above)
Comment (J3FALL) [reply fedot/core/models/evaluation/evaluation.py:221]: А, ну да, логично )
Я чето тупанул, извиняюсь :)

File: fedot/core/models/evaluation/evaluation.py (hunk 4 above)
Comment (J3FALL) [reply fedot/core/models/evaluation/evaluation.py:24]: В ветке с классификацией картинок, видимо, нужно будет это учесть.

Comment (nicl-nno) [reply fedot/core/models/evaluation/evaluation.py:24]: Ага. @v1docq учти, когда будешь дорабатывать.

File: fedot/core/composer/chain.py (hunk 1 above)
Comment (nicl-nno) [reply fedot/core/composer/chain.py]: 1. [Указал](https://github.com/nccr-itmo/FEDOT/pull/214/files#diff-f531f6aebd864a320d78a7369d86b8d8fd330665fced6295419574ff168664dbR94-R98).
2. Сейчас для бинарной классификации выдается только вероятность позитивного класса (т.к. 2-я вероятность просто дополняет её до единицы и не является полезным признаком). С full_probs в любом случае для N классов возвращается N вероятностей.

File: fedot/core/composer/node.py (hunk 2 above)
Comment (nicl-nno) [reply fedot/core/composer/node.py:122]: Done

File: test/tasks/test_classification.py (hunk 5 above)
Comment (nicl-nno) [reply test/tasks/test_classification.py:122]: Это вариант по умолчанию, он и так везде тестируется.
Добавил в этот же тест [проверку](https://github.com/nccr-itmo/FEDOT/pull/214/files#diff-cf130fc9a19e73c2b80c06f4bad94fd7fc72c80998e572a314506b6cf40b171fR117-R121), что с ним всё ок.

**Comment by nicl-nno** [comment]
+ Переструктурировал проект, добавив папка chains и data.
+ Удалил dummy-composer #120

**Review by J3FALL — APPROVED** [review]

**Comment by J3FALL** [comment]
Думаю, package в тестах тоже стоит переименовать, чтобы он соответствовал fedot/core/chains.
image
